src/main/java/com/zendesk/maxwell/producer/MaxwellNatsProducer.java

```diff
@@ -0,0 +1,242 @@
+		} catch (Exception e) {
+			this.context.terminate(new RuntimeException(e));
+		}
+		Thread thread = new Thread(this.worker, "maxwell-nats-worker");
+		thread.setDaemon(true);
+		thread.start();
+	}
+
+	@Override
+	public void push(RowMap r) throws Exception {
+		this.queue.put(r);
+	}
+}
+
+class MaxwellNatsProducerWorker extends AbstractAsyncProducer implements Runnable {
+	static final Logger LOGGER = LoggerFactory.getLogger(MaxwellNatsProducer.class);
+
+	private final String natsSubject;
+	private final int natsMaxAttempts;
+
+	private Connection natsConnection;
+	private NatsConnectionListener connectionListener;
+	private StreamingConnection streamingConnection;
+	private final ArrayBlockingQueue<RowMap> queue;
+
+
+	public MaxwellNatsProducerWorker(MaxwellContext context, String natsSubject, String clusterId,
+					 String clientId, ArrayBlockingQueue<RowMap> queue) throws IOException, InterruptedException {
+		super(context);
+
+		this.natsSubject = natsSubject;
+		this.queue = queue;
+		this.natsMaxAttempts = context.getConfig().natsMaxAttempts;
+		this.connectionListener = new NatsConnectionListener(context);
+
+
+		String natsUrl = context.getConfig().natsUrl;
+		Long pubAckTimeout = context.getConfig().natsPubAckTimeout;
+
+		Options.Builder clientOptions = new Options.Builder();
+		io.nats.streaming.Options.Builder streamingOptions = new io.nats.streaming.Options.Builder();
+
+		clientOptions
+			.server(natsUrl)
+			.connectionListener(connectionListener);
+
+		if (context.getConfig().natsMaxReconnects > 0) {
+			clientOptions.maxReconnects(context.getConfig().natsMaxReconnects);
+		}
+
+		if (context.getConfig().natsReconnectWait != null) {
+			clientOptions.reconnectWait(Duration.ofMillis(context.getConfig().natsReconnectWait));
+		}
+
+		this.natsConnection = Nats.connect(clientOptions.build());
+		streamingOptions.natsConn(natsConnection);
+
+		if (pubAckTimeout != null) {
+			streamingOptions.pubAckWait(Duration.ofMillis(pubAckTimeout));
+		}
+
+		this.streamingConnection = NatsStreaming.connect(clusterId, clientId, streamingOptions.build());
+
+	}
+
+	@Override
+	public void run() {
+		//Waiting for Nats connection
+		int attempts = 0;
+		try {
+			Thread.sleep(1000L);
+			while ( true ) {
+				// Irrespective of Nats is working or not, heartbeat should be pushed
+				if (connectionListener.isNatsRunning()){
+					attempts = 0;
+					RowMap row = queue.take();
+					this.push(row);
+				} else {
+					if (queue.peek() instanceof HeartbeatRowMap) {
+						RowMap row = queue.take();
+						this.push(row);
+					}
+
+					if (attempts < natsMaxAttempts) {
+						long sleepTime = (long) (Math.pow(2, attempts++) * 1000);
+						Thread.sleep(sleepTime);
+					} else {
+						this.context.terminate(new RuntimeException("Nats Disconnected"));
+						return;
+					}
+				}
+
+			}
+		} catch ( Exception e ) {
+			LOGGER.info("Error: ", e);
+		}
```

Comment:
can you look at `MaxwellKafkaProducer` for an example on how to handle errors, as well as how to exit gracefully (implement StoppableTask)

Reply:
While using the StoppableTask, when Nats sent Reconnected event the task wasn't able to restart and push to Nats.

Reply:
that doesn't make much sense to me.  StoppableTask really is just an interface that lets the rest of maxwell shutdown a producer/other task when it's time to exit.